blockchain/emulator.go

```diff
@@ -0,0 +1,301 @@
+/*
+ * Flow Playground
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package blockchain
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/dapperlabs/flow-playground-api/model"
+
+	"github.com/getsentry/sentry-go"
+	"github.com/onflow/cadence"
+	jsoncdc "github.com/onflow/cadence/encoding/json"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/parser"
+	emu "github.com/onflow/flow-emulator"
+	"github.com/onflow/flow-emulator/storage/memstore"
+	"github.com/onflow/flow-emulator/types"
+	flowsdk "github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/crypto"
+	"github.com/onflow/flow-go-sdk/templates"
+	"github.com/pkg/errors"
+)
+
+// blockchain interface defines an abstract API for communication with the blockchain. It hides complexity from the
+// consumer and communicates using flow native types.
+type blockchain interface {
```

Comment:
I'll leave this up to you, but I'd consider removing this interface. The rest of this package is already tightly-coupled to the concrete emulator instance and there are no consumers of this interface, so it just feels like a bit of a premature abstraction that may cause confusion for newcomers to the codebase.

Reply:
I changed for the interface to be consumed by the project. I want to keep it this way since I don't want to expose more out of the blockchain than needed, this is almost as a intended limitation to not expose too much and make the playground deal with blockchain implementation details. I also soon plan to mock it out in project to do some edge case testing.